Add homework-01

tamarinvs19 opened the pull request.

tamarinvs19 committed:
Add homework-01

sreznick commented on homework-01/database.go:
```diff
@@ -0,0 +1,43 @@
+
+func addUrl(db *sql.DB, longUrl string) (string, error) {
+	var id uint64
+	err := db.QueryRow(getIdSQL, longUrl).Scan(&id)
```
Здесь гонка. Пришли два клиента, одновременно выполнили `select`, увидели, что нет длинного, пошли создавать короткий.

tamarinvs19 replied:
Добавил mutex, по идее, теперь такого не должно быть

sreznick commented on homework-01/database.go:
```diff
@@ -0,0 +1,43 @@
+	return err
+}
+
+func addUrl(db *sql.DB, longUrl string) (string, error) {
```
db здесь делать нечего. 

По-хорошему, здесь должен быть уровень контроллера, ниже сервис, еще ниже DAO. И db должно быть инкарсулировано в DAO.

tamarinvs19 replied:
Разделил код, теперь здесь только дергаются ручки, которые уже работают с базой данных

sreznick commented on homework-01/stress_testing/stress.go:
```diff
@@ -0,0 +1,142 @@
+		"https://gitlab.com",
+	}
+
+	for i := 0; i < createResponseCount; i++ {
```
У вас получится 10 реальных созданий и 9990 обнаружений уже созданных.

То есть тест почти всегда ходит не по основному пути.

tamarinvs19 replied:
Теперь добавляю суффикс к этим адресам, большинство запросов будут уникальными

sreznick commented on homework-01/stress_testing/stress.go:
```diff
@@ -0,0 +1,142 @@
+	fmt.Println("Start success get requests")
+	tinyUrl := requestCreate("https://emkn.ru")
+	for i := 0; i < getSuccessResponseCount; i++ {
+		code := requestGet(tinyUrl)
```
Очень хорошо, что отключили редирект. Но совсем бы здорово еще на многих url проверить. Один может где-то закешироваться и дать неточную картину.

tamarinvs19 replied:
Теперь здесь много рандомных адресов

sreznick commented on homework-01/database.go:
```diff
@@ -0,0 +1,43 @@
+func addUrl(db *sql.DB, longUrl string) (string, error) {
+	var id uint64
+	err := db.QueryRow(getIdSQL, longUrl).Scan(&id)
+	if err != nil {
```
Кажестя, ошибки могут быть разными.

Это не обязательно `ErrNoRows` в результате `Scan`

tamarinvs19 replied:
Добавил обработку ErrNoRows отдельно

sreznick commented on homework-01/stress_testing/stress.go:
```diff
@@ -0,0 +1,142 @@
+
+func runGetInvalid() {
+	fmt.Println("Start bad get requests")
+	badTinyUrl := "_______"
```
Неинтересный тест. Оно породит 404 не дойдя до базы. 

Надо бы поинтереснее и поразнообразнее.

tamarinvs19 replied:
Добавил больше простых тестов и много рандомных, но, чисто теоретически, они могут работать неправильно, так как я не могу гарантировать, что сгенерированный tinyUrl не будет существовать. Сделал так, чтобы ко всем добавлялся 'z' в конец, так это будут самые большие id, соответственно, вероятность того, что мы уже дошли до них очень низка.

sreznick reviewed (commented):
-1 за гонку и неаккуратную проверку ошибки

-2 за недостатки тестов

tamarinvs19 committed:
Refactoring and bug fix:

1. Added mutex in creation of the new tiny url
2. Project structure was separated to DB, logic, controller and router
3. Created singleton for DB
4. Update tests

tamarinvs19 committed:
upd
